src/tracers.jl

```diff
@@ -1,194 +1,184 @@
 abstract type AbstractTracer <: Number end
 
+# We represent sparse Hessians as sets of index tuples.
+const AbstractPairSet{T} = AbstractSet{Tuple{T,T}}
+
 # Convenience constructor for empty tracers
 empty(tracer::T) where {T<:AbstractTracer} = empty(T)
+empty(T) = T()
 
-const SET_TYPE_MESSAGE = """
-The provided index set type `S` has to satisfy the following conditions:
+sparse_vector(T, index) = T([index])
 
-- it is an iterable with `<:Integer` element type
-- it implements `union`
+#================#
+# Elementwise OR #
+#================#
 
-Subtypes of `AbstractSet{<:Integer}` are a natural choice, like `BitSet` or `Set{UInt64}`.
-"""
+×(a::G, b::G) where {G<:AbstractSet} = Set((i, j) for i in a, j in b)
 
 #==============#
 # Connectivity #
 #==============#
 
 """
-    ConnectivityTracer{I,S}(indexset) <: Number
+    ConnectivityTracer{C}(indexset) <: Number
 
 Number type keeping track of input indices of previous computations.
 
-$SET_TYPE_MESSAGE
-
 For a higher-level interface, refer to [`connectivity_pattern`](@ref).
+
+```jldoctest
+julia> inputs = Set([1, 3])
+Set{Int64} with 2 elements:
+  3
+  1
+
+julia> ConnectivityTracer(inputs)
+ConnectivityTracer{Set{Int64}}(1, 3)
+```
 """
-struct ConnectivityTracer{I<:Integer,S} <: AbstractTracer
-    inputs::S # indices of connected, enumerated inputs
-end
-function ConnectivityTracer(inputs::S) where {S}
-    I = eltype(S)
-    return ConnectivityTracer{I,S}(inputs)
+struct ConnectivityTracer{C<:AbstractSet{<:Integer}} <: AbstractTracer
+    inputs::C # sparse binary vector representing non-zero indices of connected, enumerated inputs
 end
 
-function Base.show(io::IO, t::ConnectivityTracer{I,S}) where {I,S}
+function Base.show(io::IO, t::ConnectivityTracer)
     return Base.show_delim_array(
-        io, convert.(Int, inputs(t)), "ConnectivityTracer{$I,$S}(", ',', ')', true
+        io, convert.(Int, sort!(collect(t.inputs))), "$(typeof(t))(", ',', ')', true
```

Comment:
won't this modify the object in-place when `collect` does nothing?

Reply:
Nice catch. It probably does.